File: packages/capacitor-notificare-loyalty/ios/Plugin/NotificareLoyaltyPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareLoyaltyKit

@objc(NotificareLoyaltyPlugin)
public class NotificareLoyaltyPlugin: CAPPlugin {
    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }
    
    @objc func fetchPassBySerial(_ call: CAPPluginCall) {
        guard let serial = call.getString("serial") else {
            call.reject("Missing 'serial' parameter.")
            return
        }
        
        Notificare.shared.loyalty().fetchPass(serial: serial) { result in
            switch result {
            case let .success(pass):
                do {
                    call.resolve([
                        "result": try pass.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func fetchPassByBarcode(_ call: CAPPluginCall) {
        guard let barcode = call.getString("barcode") else {
            call.reject("Missing 'barcode' parameter.")
            return
        }
        
        Notificare.shared.loyalty().fetchPass(barcode: barcode) { result in
            switch result {
            case let .success(pass):
                do {
                    call.resolve([
                        "result": try pass.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func present(_ call: CAPPluginCall) {
        guard let json = call.getObject("pass") else {
            call.reject("Missing 'pass' parameter.")
            return
        }
        
        let pass: NotificarePass
        
        do {
            pass = try NotificarePass.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot present a pass with a nil root view controller.")
                return
            }
            
            Notificare.shared.loyalty().present(pass: pass, in: rootViewController)
        }
    }
}

File: packages/capacitor-notificare-loyalty/android/src/main/java/re/notifica/loyalty/capacitor/NotificareLoyaltyPlugin.kt
package re.notifica.loyalty.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.loyalty.ktx.loyalty
import re.notifica.loyalty.models.NotificarePass

@CapacitorPlugin(name = "NotificareLoyaltyPlugin")
public class NotificareLoyaltyPlugin : Plugin() {
    @PluginMethod
    public fun fetchPassBySerial(call: PluginCall) {
        val serial = call.getString("serial") ?: run {
            call.reject("Missing 'serial' parameter.")
            return
        }

        Notificare.loyalty().fetchPassBySerial(serial, object : NotificareCallback<NotificarePass> {
            override fun onSuccess(result: NotificarePass) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchPassByBarcode(call: PluginCall) {
        val barcode = call.getString("barcode") ?: run {
            call.reject("Missing 'barcode' parameter.")
            return
        }

        Notificare.loyalty().fetchPassByBarcode(barcode, object : NotificareCallback<NotificarePass> {
            override fun onSuccess(result: NotificarePass) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun present(call: PluginCall) {
        val json = call.getObject("pass") ?: run {
            call.reject("Missing 'pass' parameter.")
            return
        }

        val pass: NotificarePass = try {
            NotificarePass.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        val activity = activity ?: run {
            call.reject("Cannot present a pass without an activity.")
            return
        }

        Notificare.loyalty().present(activity, pass)
        call.resolve()
    }
}
